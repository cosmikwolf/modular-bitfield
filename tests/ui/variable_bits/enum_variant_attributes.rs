use modular_bitfield::prelude::*;

// Test: Explicit bits on enum variants
#[derive(Specifier)]
#[bits = (8, 16, 24)]
pub enum ExplicitBitsOnVariants {
    #[bits = 8]  // Should match position 0
    A(u8),
    #[bits = 16] // Should match position 1
    B(u16),
    #[bits = 24] // Should match position 2
    C(B24),
}

// Test: Conflicting explicit bits
#[derive(Specifier)]
#[bits = (8, 16, 24)]
pub enum ConflictingBits {
    #[bits = 8]  // OK
    A(u8),
    #[bits = 32] // Conflicts with tuple position 1 which specifies 16
    B(u16),
    C(B24),
}

// Test: bits attribute on unit variant
#[derive(Specifier)]
#[bits = (8, 16)]
pub enum BitsOnUnitVariant {
    #[bits = 8]
    A,  // Unit variants shouldn't have bits attribute
    B(u16),
}

// Test: Invalid bits attribute format
#[derive(Specifier)]
pub enum InvalidBitsFormat {
    #[bits = "8"]  // Should be numeric literal
    A(u8),
    #[bits(8)]     // Should be #[bits = N]
    B(u16),
}

// Test: discriminant on data variant without discriminant_bits
#[derive(Specifier)]
#[bits = (8, 16)]
pub enum DiscriminantWithoutDiscriminantBits {
    #[discriminant = 0]  // discriminant without discriminant_bits might not make sense
    A(u8),
    #[discriminant = 1]
    B(u16),
}

fn main() {}