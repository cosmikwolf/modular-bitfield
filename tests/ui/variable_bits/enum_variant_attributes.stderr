error: variant #[bits = 32] conflicts with tuple position 1 size 16
  --> tests/ui/variable_bits/enum_variant_attributes.rs:21:5
   |
21 | /     #[bits = 32] // Conflicts with tuple position 1 which specifies 16
22 | |     B(u16),
   | |__________^

error: enums with data variants require explicit #[bits = N] attribute
  --> tests/ui/variable_bits/enum_variant_attributes.rs:37:1
   |
37 | / pub enum InvalidBitsFormat {
38 | |     #[bits = "8"]  // Should be numeric literal
39 | |     A(u8),
40 | |     #[bits(8)]     // Should be #[bits = N]
41 | |     B(u16),
42 | | }
   | |_^

error: attribute value must be a literal
 --> tests/ui/variable_bits/enum_variant_attributes.rs:5:10
  |
5 | #[bits = (8, 16, 24)]
  |          ^^^^^^^^^^^

error: attribute value must be a literal
  --> tests/ui/variable_bits/enum_variant_attributes.rs:17:10
   |
17 | #[bits = (8, 16, 24)]
   |          ^^^^^^^^^^^

error: attribute value must be a literal
  --> tests/ui/variable_bits/enum_variant_attributes.rs:28:10
   |
28 | #[bits = (8, 16)]
   |          ^^^^^^^

error: attribute value must be a literal
  --> tests/ui/variable_bits/enum_variant_attributes.rs:46:10
   |
46 | #[bits = (8, 16)]
   |          ^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/variable_bits/enum_variant_attributes.rs:11:5
   |
11 | /     #[bits = 24] // Should match position 2
12 | |     C(B24),
   | |          ^
   | |          |
   | |__________expected `u32`, found `B24`
   |            arguments to this function are incorrect
   |
note: associated function defined here
  --> src/lib.rs
   |
   |     fn into_bytes(input: Self::InOut) -> Result<Self::Bytes, OutOfBounds>;
   |        ^^^^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/variable_bits/enum_variant_attributes.rs:4:10
   |
4  | #[derive(Specifier)]
   |          ^^^^^^^^^
   |          |
   |          expected `B24`, found `u32`
   |          arguments to this enum variant are incorrect
   |
note: tuple variant defined here
  --> tests/ui/variable_bits/enum_variant_attributes.rs:12:5
   |
12 |     C(B24),
   |     ^
   = note: this error originates in the derive macro `Specifier` (in Nightly builds, run with -Z macro-backtrace for more info)
